File: py/algorithms/postOrderDFS_naryTree.py
def postOrderDFS_nary(node):
    """
    This algorithm traverses an N-Ary tree
    in post-order fashion recursively. 

    Assumes input nodes have a property .children
    containing all the children the node is connected to,
    and a property .val
    """
    if not node:
        return
    output = []
    postOrderHelper_recursive(node, output)
    return output

def postOrderHelper_recursive(node, output):
    if not node:
        return
    for child in node.children:
        postOrderHelper_recursive(child, output)
    output.append(node.val)



def postorder_iterative(self, node):
    """
    This algorithm traverses an N-Ary tree in post-order fashion iteratively. 

    Assumes input nodes have a property .children containing all the children the node is connected to,
    and a property .val.

    Key thing: emulate the recursive call stack using an actual stack.
    How we convert recursive code to iterative code. 
    """
    if not node:
        return
    stack = [(node,0)]
    output = [] 
    while stack:
        node, currIdxChild = stack.pop() 
        if currIdxChild == len(node.children):
            output.append(node.val)
        else:
            stack.append((node, currIdxChild+1))
            stack.append((node.children[currIdxChild], 0))
            
    return output 

File: ts/src/algorithms/post_order_dfs_nary_tree.ts
import { NaryTreeNode } from 'src/utils/binary_tree';

function postOrderHelperRecursive<T>(node: NaryTreeNode<T>, output: T[]): void {
    node.children.forEach((child) => {
        postOrderHelperRecursive(child, output);
    });
    output.push(node.val);
}

/**
 *  This code describes the recursive post order traversal
    of a n-ary tree. Before any given node in the tree is visited,
    all the nodes in the left subtree of the current node, and all
    the nodes in the right subtree of the current node are visited.

 * @param {NaryTreeNode} node Root of Nary Tree
 * @returns {number[]} Output values of n-ary tree in post order 
 */
function postOrderRecursive<T>(node: NaryTreeNode<T>): T[] {
    const output: T[] = [];
    postOrderHelperRecursive<T>(node, output);
    return output;
}

/**
 *  This code describes the iterative post order traversal
    of a n-ary tree. Before any given node in the tree is visited,
    all the nodes in the left subtree of the current node, and all
    the nodes in the right subtree of the current node are visited.

 * @param {NaryTreeNode} node Root of Nary Tree
 * @returns {number[]} Output values in binary tree in post order 
 */
function postOrderIterative<T>(node: NaryTreeNode<T>): T[] {
    const stack = [[node, 0]];
    const output: T[] = [];
    while (stack.length > 0) {
        const [currNode, idx] = stack.pop() as [NaryTreeNode<T>, number];
        if (idx === currNode.children.length) {
            output.push(currNode.val);
        } else {
            stack.push([currNode, idx + 1]);
            stack.push([currNode.children[idx], 0]);
        }
    }
    return output;
}

export { postOrderRecursive, postOrderIterative };
